### ISSUE TITLE
Improvements to streaming JSON consumption

### ISSUE THREAD

**Issue by UberMouse** [koordinates/kart#412]
Our current JSON format is not ideal for consumption in a streaming manner. Parsing a stream of JSON is inherently more complex than parsing a full chunk of JSON. However, I don't see any specific reason that we need to stick with emitting a single JSON object in a streaming format. 

Conceptually all a diff really represents is a series of feature change objects representing the changes to the features since the previous diff target. These could be emitted one at a time using [JSONL](   https://jsonlines.org) which simplifies the parsing of the JSON stream as you no longer need to parse the stream itself, just parse each individual JSON object on each line. 

That means you no longer need to use a streaming aware JSON parser and can either parse each emitted chunk on stdout or have a very simple parser that looks for the linefeed indicating the JSONL is over and parsing that line; depending on whether it's guaranteed that each write to stdout is a full JSONL object or not (not sure if that's the case unless you flush after every write? which might have performance implications?).

This also means we would need to change the format of the emitted JSON, since it's no longer going to be a single object, but a series of individual JSON objects. Currently, in the Koordinates App we transform the features emitted by Sno to a different format for ease of consumption, at the same time as this change to the feature change format we could make it more closely resemble the transformed object so we can skip that step.

In which case, I would imagine something like this for the format of each line in the JSONL stream. 

```
{
  "dataset": "datasetId",
  "type": "update", // one of "update", "insert", "delete"
  "feature": {
    "id": 45, // value of the primary key
    "geom": { // can be null if there is no geometry for the feature
      "state": "changed", // "unchanged" or "changed"
      "value": "{\"type\":\"Point\",\"coordinates\":[174.8680551667,-41.0570918167]}",
      "oldValue": null
    },
    "properties": {
      "fid": {
        "state": "unchanged",
        "value": 45
      },
    },
	"meta": {
		// same format as properties
	}
  }
}

```

Which is basically a property to associate it with the dataset it's part of, what has happened to the feature (created, deleted or updated) and then information about the feature and its properties. Each of  the properties contains information about what change has happened to it (created or deleted and then changed/unchanged depending on what type of change caused the feature to be present in the diff).

We need to do that sort of post-processing on the features for the Koordinates App so we can display at the feature level which properties have changed for a given changed feature, since we only want to highlight the ones that have actually changed. It seems like a reasonable way of structuring a changed feature though so feels like a reasonable structure for sno to output over the current `+`/`-` format. But happy to discuss alternatives that could meet the Koordinates App needs and Snos needs better.

**Comment by craigds** [comment]
* JSONL: 👍 , we should definitely do this.

* GeoJSON geometries: We used them in the very early days, but replaced with hexwkb for size and performance reasons (#71.) Worth revisiting? If we did revisit, I don't think I'd want them double-escaped like that - they should probably be properly nested JSON objects ideally. But GeoJSON lacks a few key things like curve geometry support, as well as being really verbose...

* geometries separate to 'properties': not sure - is this useful? Means you lose the name of the geometry field unless you inspect the schema. i guess the reverse argument is that _at present_ you don't know _which_ field is the geometry field unless you inspect the schema. I've got no real opinion on which approach is better.

* Per-field diffs: I don't have any real arguments for or against, but it _looks_ quite noisy to me. Perhaps it's useful, but it's also probably quite easy for clients to calculate

* `meta`: Perhaps that was meant to be a top-level key? i.e. each change in the stream would either have a `meta` or `feature` key, but not both at once.

* `id` stored separately: looks good - though there is a problem - an update can change the primary key value for a feature, so it's ambiguous here whether we'd display the old or new PK. Of course if we just picked one, then that'd be fine I guess, as long as the old PK is available in `properties`.

I guess I'd be a bit concerned about having `json` and `jsonl` output formats differ by too much, just for sanity's sake. Obviously the top level structure has to differ, but things like the per-field diffs don't really _need_ to differ.

**Comment by UberMouse** [comment]
> GeoJSON geometries:
I dumped one a test sno diff run through our snoFeatureToFeatureChange converter to a file for that example so that's why the GeoJSON string is double escaped. If we could find a format for sno to output that OpenLayers is capable of consuming that would get rid of the requirement to convert it to GeoJSON. OpenLayers doesn't support HexWKB.

> geometries separate to properties:
Yea, not knowing which property is the geometry field is why it's promoted to its own property. An alternative solution could be thought up if required.

> Per-field diffs:
Having sno emit in this format is just a performance optimization so we don't have to retransform the JSON as it comes into the app. We need this information to be able to display the changes properly in the diff table

> meta:
:shrug: we totally ignore anything to do with metadata in the app at the moment so I don't really know anything about it. I just slapped that in to the example

> id:
Yea, open to suggestions there. I suppose we could apply the same changed property format to it. I'm not sure the app correctly handles diffs with changes to the primary key or not currently.

Yea, we're just sort of discussing two things in the issue

1. Adding a JSONL format
2. Having that format better match what the Koordinates App needs for a performance optimization

1. is the primary thing, 2. is just nice to have, but we don't currently have that and once we have everything moved to a streaming pipeline the cost of retransforming the JSON is less problematic so it's not a big concern to me.

The important part of the JSONL format is that it's at least something like this
`{ "dataset": "datasetId", "feature": { // whatever format the feature change is } }`
since the dataset is the only piece of information we need apart from the feature itself.

**Comment by rcoup** [comment]
> OpenLayers doesn't support HexWKB.

We could write that, it's not _particularly_ complicated.

**Comment by rcoup** [comment]
Ignoring things like schema/meta changes, which fields are geometries, pk changes, etc are all only going to be temporary, so feels like this needs to be a message stream? Then we can add different messages to the protocol as needed.

```
{"datasetId": "outlines", "type": "meta-schema", "+": {...}, "-": {...}}  # old & new field schemas for referencing in feature
{"datasetId": "outlines", "type": "meta-change", "id": "schema.json", "+": [...], "-": [...]}
{"datasetId": "outlines", "type": "meta-change", "id": "schema.json", "+": [...], "-": [...]}
{"datasetId": "outlines", "type": "meta-change", "id": "title", "+": "New Outlines", "-": "Old Outlines"}
{"datasetId": "outlines", "type": "feature-change", "+": {...}, "-": {...}}  # update
{"datasetId": "outlines", "type": "feature-change", "+": {...}}  # insert
{"datasetId": "outlines", "type": "feature-change", "-": {...}}  # delete
```

**Comment by rcoup** [comment]
> We need to do that sort of post-processing on the features for the Koordinates App so we can display at the feature level which properties have changed for a given changed feature, since we only want to highlight the ones that have actually changed. It seems like a reasonable way of structuring a changed feature though so feels like a reasonable structure for sno to output over the current +/- format.

I don't understand why the diff needs processed to a _new structure_ to populate a table/map essentially, but I'm probably missing something? This data is by definition _never changing_, so it doesn't need to be preserved for any other purpose AFAICT. It's not a model we need to maintain as state, or that is editable anywhere. I'd imagine the process would go something like this:

1. setup table field headers/data formatting/etc from schema
2. for each feature, figure out whether it's inserted/deleted/updated
3. for each field, figure out whether it's changed
4. add row to table
5. add feature to map
6. back to (2) for the next feature

**Comment by UberMouse** [comment]
>so feels like this needs to be a message stream?
I'm not really following why this is a message stream?

> don't understand why the diff needs processed to a new structure to populate a table/map essentially
This preprocessing is doing step 2/3 from your example. If we didn't preprocess it we're just doing the same work at render time. So it's either

1. pay the cost upfront and have a simple model exposed to the UI
or
2. pay the cost during the rendering phase and require the UI to calculate the same information

**Comment by rcoup** [comment]
> I'm not really following why this is a message stream?

The JSON objects aren't homogenous, there are different message types (`"type": ...`), and some should logically come before others in the stream, and the client needs to act on them in different ways.

**Comment by UberMouse** [comment]
What are the different message types though? I'm not understanding what we need other than a series of changed features?

I guess we need something else for meta/schema changes because they don't belong to features?

**Comment by rcoup** [comment]
> So it's either
> 
> pay the cost upfront and have a simple model exposed to the UI
> or
> pay the cost during the rendering phase and require the UI to calculate the same information

Sure, but we only need to do it once regardless? I guess my thinking is:
1. takes the same time to evaluate it regardless of where it happens
2. copying a 100MB diff's worth of objects into a different structure is another 100MB of memory/allocations/gc/etc
3. which doesn't need to happen if you figure out and act on it in the same place

hmmm ... wrt (1) _is_ there a practical difference where it happens — in terms of UI/interactivity/event-loops/etc?

**Comment by UberMouse** [comment]
We're only doing it once regardless though? As features appear in stream -> transform them. We're not paying extra memory because the initial object is discarded and garbage collected (so higher garbage collection pressure yes, but not higher memory pressure).

Yes, there is a practical difference in terms of where it happens. If we do it during the render of the table it's going to slow that render down.

**Comment by rcoup** [comment]
> I guess we need something else for meta/schema changes because they don't belong to features?

Data types for fields; which fields are PKs & geometries; schema changes... and any other relevant changes which need displayed (CRS/title/description/metadata/license...)

**Comment by UberMouse** [comment]
Ok, so we're talking about a superset of the existing diff output. That makes sense then.

**Comment by UberMouse** [comment]
Any further thoughts on this? I've figured out how to associate changed features with the dataset they belong to with stream-json so I can keep trucking along with the current format. 

But I'm fairly certain at some point we will want to move to a format like this as it's quite a bit more flexible.

**Comment by olsen232** [comment]
This is mostly done as `kart show -output-format=json-lines` and the related commands. This solves the problem of "too hard to parse while streaming". The stream now also contains information about the schema, regardless of whether it has changed or not - this means that the primary key(s) and geometry column(s) are straight-forward to identify.

The main thing suggested above that is not done is changing the output format so that the app doesn't have to do any postprocessing - eg geojson instead of hexwkb.

**Comment by craigds** [comment]
I'll make a call and say: the important bits of this ticket are solved by `kart diff --output-format=json-lines` so I'm going to close the issue.

If someone thinks geojson geometries in diffs would be very useful to them for some reason, please open a new issue for that :)
